ygen/protogen.go

```diff
@@ -292,40 +296,162 @@ type protoMsgConfig struct {
 //  name of the package it is within, the code for the message, and any imports for packages that are referenced by
 //  the message.
 func writeProto3Msg(msg *yangDirectory, msgs map[string]*yangDirectory, state *genState, cfg protoMsgConfig) (*generatedProto3Message, []error) {
-	var pkg string
+	if cfg.nestedMessages {
+		if !msg.isFakeRoot && len(msg.path) != 3 {
+			// We only operate on messages that are length 3 (i.e., are children of a
+			// module). The path is in the format []{"", <module>, <element>} hence we
+			// match len == 3.
+			return nil, nil
+		}
+		return writeProto3MsgNested(msg, msgs, state, cfg)
+	}
+	return writeProto3MsgSingleMsg(msg, msgs, state, cfg)
+}
+
+// writeProto3MsgNested returns a nested set of protobuf messages for the message
+// supplied, which is expected to be a top-level message that code generation is
+// being performed for. It takes:
+// 	- msg: the top-level directory definition
+// 	- msgs: the set of message definitions (keyed by path) that are to be output
+//	- state: the current code generation state.
+//  - cfg: the configuration for the current code generation.
+// It returns a generated protobuf3 message.
+func writeProto3MsgNested(msg *yangDirectory, msgs map[string]*yangDirectory, state *genState, cfg protoMsgConfig) (*generatedProto3Message, []error) {
+	var gerrs []error
+	childMsgs := []*generatedProto3Message{}
+	// Find all the children of the current message that should be output.
+	for _, n := range msgs {
+		if isDirectEntryChild(msg.entry, n.entry, cfg.compressPaths) {
+			cmsg, errs := writeProto3MsgNested(n, msgs, state, cfg)
+			if errs != nil {
+				gerrs = append(gerrs, errs...)
+				continue
+			}
+			childMsgs = append(childMsgs, cmsg)
+		}
+	}
+
+	pkg, err := protobufPackageForMsg(msg, state, cfg.compressPaths)
+	if err != nil {
+		return nil, append(gerrs, err)
+	}
+
+	// Generate this message, and its associated messages.
+	msgDefs, errs := genProto3Msg(msg, msgs, state, cfg, pkg, childMsgs)
+	if errs != nil {
+		return nil, append(gerrs, errs...)
+	}
+
+	gmsg, errs := genProto3MsgCode(pkg, msgDefs)
+	if errs != nil {
+		return nil, append(gerrs, errs...)
+	}
+
+	// Inherit the set of imports that are required for this child. We
+	// skip any that are relative imports as these are only needed for
+	// the case that we have different files per hierarchy level and
+	// are not nesting messages.
+	imports := []string{}
+	if msg.isFakeRoot {
+		imports = gmsg.RequiredImports
+	} else {
+		allImports := map[string]bool{}
```

Comment:
We can't do `var allImports map[string]bool{}` since this will mean that @364 here, we'll get a panic since we assign to a nil map. Fixed elsewhere though.